File: Kotlin/src/main/kotlin/com/gildedrose/GildedRose.kt
package com.gildedrose

class GildedRose(var items: List<Item>) {

    fun updateQuality() {
        for (i in items.indices) {
            when(items[i].name) {
                "Aged Brie" -> agedBrie(items[i])
                "Backstage passes to a TAFKAL80ETC concert" -> backstagePasses(items[i])
                "Sulfuras, Hand of Ragnaros" -> continue
                else -> {
                    regularItem(items[i])
                }
            }
        }
    }

    fun regularItem(item: Item) {
        item.sellIn--
        if (item.sellIn < 0) {
            qualityMoreThan0(item)
        }
        qualityMoreThan0(item)
    }

    fun agedBrie(item: Item) {
        item.sellIn--
        if (item.sellIn < 0) {
            qualityLessThan50(item)
        }
        qualityLessThan50(item)
    }

    fun backstagePasses(item: Item) {
        item.sellIn--
        if(item.sellIn < 0) {
            item.quality = 0
            return
        }
        if(item.sellIn < 5) {
            qualityLessThan50(item)
        }
        if(item.sellIn < 10) {
            qualityLessThan50(item)
        }
        qualityLessThan50(item)
    }

    fun qualityLessThan50(item: Item) {
        if(item.quality < 50) {
            item.quality++
        }
    }

    fun qualityMoreThan0(item: Item) {
        if(item.quality > 0) {
            item.quality--
        }
    }

}

File: TypeScript/app/gilded-rose.ts
export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  updateQuality() {
    for (const currentItem of this.items) {
      if (currentItem.name === 'Aged Brie' || currentItem.name.startsWith('Backstage pass')) {
        currentItem.quality++;
        if (currentItem.name.startsWith('Backstage pass')) {
          if (currentItem.sellIn <= 10) {
            currentItem.quality += currentItem.sellIn <= 5 ? 2 : 1;
          }
        }
      } else {
        if (currentItem.quality > 0) {
          if (currentItem.name != 'Sulfuras, Hand of Ragnaros') {
            currentItem.quality = currentItem.quality - 1;
          }
        }
      }
      if (currentItem.name != 'Sulfuras, Hand of Ragnaros') {
        currentItem.sellIn = currentItem.sellIn - 1;
      }
      if (currentItem.sellIn < 0) {
        if (currentItem.name != 'Aged Brie') {
          if (!currentItem.name.startsWith('Backstage pass')) {
            if (currentItem.quality > 0) {
              if (currentItem.name != 'Sulfuras, Hand of Ragnaros') {
                currentItem.quality = currentItem.quality - 1;
              }
            }
          } else {
            currentItem.quality = currentItem.quality - currentItem.quality
          }
        } else {
          if (currentItem.quality < 50) {
            currentItem.quality = currentItem.quality + 1;
          }
        }
      }

      if (currentItem.name !== 'Sulfuras, Hand of Ragnaros') {
        currentItem.quality = Math.min(currentItem.quality, 50);
      }
    };

    return this.items;
  }
}
